src/plugins/plugins_ex.h

```diff
@@ -386,119 +333,133 @@ struct breakpoint_by_pid_searcher
                                   hook_cb_t hook_cb,
                                   IB init_breakpoint,
-                                  const char* trap_name = nullptr,
-                                  AP ap = allocator<Params>(),
-                                  AT at = allocator<drakvuf_trap_t>())
+                                  const char* trap_name = nullptr)
     {
-        std::unique_ptr<Params> params;
-        std::unique_ptr<drakvuf_trap_t> trap;
-        if (!init_memory(plugin, trap, params, hook_cb, trap_name, ap, at))
-            return nullptr;
+        using Plugin = std::remove_pointer<decltype(this)>::type;
+
+        auto trap = new drakvuf_trap_t;
+
+        if constexpr (std::is_same_v<Params, void>) {
+            trap->data = new plugin_data<Plugin, Params>(this, nullptr);
+        } else {
+            trap->data = new plugin_data<Plugin, Params>(this, new Params);
+        }
+
+        trap->name = trap_name;
+        trap->cb = hook_cb;
+        trap->type = BREAKPOINT;
 
-        if (!init_breakpoint(drakvuf, info, trap.get()))
+        if (!init_breakpoint(drakvuf, info, trap))
         {
             PRINT_DEBUG("%s for %s\n", ERROR_MSG_ADDING_TRAP, trap_name ? trap_name : trap->name);
             return nullptr;
         }
 
-        attach_plugin_params(trap.get());
-        params.release();
-        return trap.release();
+        traps.push_back(std::move(trap));
+        return traps.back();
     }
 
-    virtual void destroy_trap(drakvuf_t drakvuf, drakvuf_trap_t* trap);
-    drakvuf_trap_t* detach_plugin_params(drakvuf_trap_t* data);
-    static void destroy_plugin_params(drakvuf_trap_t* data);
-
-protected:
-    void attach_plugin_params(drakvuf_trap_t* data);
-
-    template<typename Plugin, typename Params, typename AP = allocator<Params>, typename AT = allocator<drakvuf_trap_t>>
-    bool init_memory(Plugin* plugin,
-                     std::unique_ptr<drakvuf_trap_t>& trap,
-                     std::unique_ptr<Params>& params,
-                     hook_cb_t hook_cb,
-                     const char* trap_name = nullptr,
-                     AP ap = allocator<Params>(),
-                     AT at = allocator<drakvuf_trap_t>())
+    template<typename Params>
+    void destroy_plugin_params(drakvuf_trap_t* target)
     {
-        trap.reset(at.allocate(1));
-        if (!trap)
-        {
-            PRINT_DEBUG("%s. Failed to allocate a memory for trap of %s\n", ERROR_MSG_ADDING_TRAP, trap_name);
-            return false;
-        }
+        using Plugin = std::remove_pointer<decltype(this)>::type;
+
+        std::for_each(traps.begin(), traps.end(), [=](const auto& trap) {
+            if (trap != target)
+                return;
+
+            auto data = static_cast<plugin_data<Plugin, Params>*>(trap->data);
+            if (data->params)
+                delete data->params;
+            else
+                PRINT_DEBUG("[PLUGINS_EX] BUG?? params already deleted before");
```

Comment:
> If the operand to the delete operator is a modifiable l-value, its value is undefined after the object is deleted.

So I think we don't really know that params were already deleted if we don't explicitly assign zero to `data->params` after `delete`-ing.